src/common/src/util/memcmp_encoding.rs

```diff
@@ -0,0 +1,558 @@
+            DataType::Float64 => size_of::<OrderedF64>(),
+            DataType::Date => size_of::<NaiveDateWrapper>(),
+            DataType::Time => size_of::<NaiveTimeWrapper>(),
+            DataType::Timestamp => size_of::<NaiveDateTimeWrapper>(),
+            DataType::Timestamptz => size_of::<i64>(),
+            DataType::Boolean => size_of::<u8>(),
+            // IntervalUnit is serialized as (i32, i32, i64)
+            DataType::Interval => size_of::<(i32, i32, i64)>(),
+            DataType::Decimal => {
+                deserializer.deserialize_decimal()?;
+                0 // the len is not used since decimal is not a fixed length type
+            }
+            // these two types is var-length and should only be determine at runtime.
+            // TODO: need some test for this case (e.g. e2e test)
+            DataType::List { .. } => deserializer.skip_bytes()?,
+            DataType::Struct(t) => t
+                .fields
+                .iter()
+                .map(|field| {
+                    // use default null tags inside composite type
+                    calculate_encoded_size_inner(
+                        field,
+                        DEFAULT_NULL_TAG_NONE,
+                        DEFAULT_NULL_TAG_SOME,
+                        deserializer,
+                    )
+                })
+                .try_fold(0, |a, b| b.map(|b| a + b))?,
+            DataType::Jsonb => deserializer.skip_bytes()?,
+            DataType::Varchar => deserializer.skip_bytes()?,
+            DataType::Bytea => deserializer.skip_bytes()?,
+        };
+
+        // consume offset of fixed_type
+        if deserializer.position() == base_position + 1 {
+            // fixed type
+            deserializer.advance(len);
+        }
+    } else {
+        return Err(memcomparable::Error::InvalidTagEncoding(null_tag as _));
+    }
+
+    Ok(deserializer.position() - base_position)
+}
+
+pub fn encode_value(value: impl ToDatumRef, order: OrderType) -> Result<Vec<u8>> {
+    let mut serializer = memcomparable::Serializer::new(vec![]);
+    serialize_datum(value, order, &mut serializer)?;
+    Ok(serializer.into_inner())
+}
+
+pub fn decode_value(ty: &DataType, encoded_value: &[u8], order: OrderType) -> Result<Datum> {
+    let mut deserializer = memcomparable::Deserializer::new(encoded_value);
+    Ok(deserialize_datum(ty, order, &mut deserializer)?)
+}
+
+pub fn encode_array(array: &ArrayImpl, order: OrderType) -> Result<Vec<Vec<u8>>> {
+    let mut data = Vec::with_capacity(array.len());
+    for datum in array.iter() {
+        data.push(encode_value(datum, order)?);
+    }
+    Ok(data)
+}
+
+/// This function is used to accelerate the comparison of tuples. It takes datachunk and
+/// user-defined order as input, yield encoded binary string with order preserved for each tuple in
+/// the datachunk.
+pub fn encode_chunk(chunk: &DataChunk, column_orders: &[ColumnOrder]) -> Vec<Vec<u8>> {
```

Comment:
further optimization: we do not need to encode the row which is not in the visibility bitmap and return an empty hole in the result.